handlers/photo-haptics-audio-handler/src/server.ts

```diff
@@ -0,0 +1,327 @@
+    if (!hasAudioHaptic && !hasText) {
+        console.warn("Segment audio-haptic renderers not supported!");
+        const response = utils.generateEmptyResponse(req.body["request_uuid"]);
+        if (ajv.validate("https://image.a11y.mcgill.ca/handler-response.schema.json", response)) {
+            res.json(response);
+        } else {
+            console.error("Failed to generate a valid empty response!");
+            console.error(ajv.errors);
+            res.status(500).json(ajv.errors);
+        }
+        return;
+    }
+
+    // *******************************************************
+    // Audio TTS
+    // *******************************************************
+    // Begin forming text...
+    // This is variable depending on which preprocessor data is available.
+    const ttsData: utils.TTSSegment[] = [];
+    ttsData.push({"value": utils.generateIntro(preSecondCat), "type": "text"});
+    if (preSemSeg) {
+        // Use all segments returned for now.
+        // Filtering may be helpful later.
+        ttsData.push(...utils.generateSemSeg(preSemSeg));
+        if (preObjDet && preGroupData) {
+            ttsData.push({"value": "It also", "type": "text"});
+        }
+    }
+    if (preObjDet && preGroupData) {
+        ttsData.push(...utils.generateObjDet(preObjDet, preGroupData));
+    }
+
+        // Concatenate adjacent text entries
+        for (let i = 0; i < ttsData.length - 1; i++) {
+            if (ttsData[i].type === "text" && ttsData[i+1].type === "text") {
+                ttsData[i].value += " " + ttsData[i+1].value;
+                ttsData.splice(i+1, 1);
+            }
+        }
+
+    // Generate rendering title
+    const renderingTitle = utils.renderingTitle(preSemSeg, preObjDet, preGroupData);
+
+    // Construct text (if requested)
+    if (hasText) {
+        const textString = ttsData.map(x => x["value"]).join(" ");
+        const rendering = {
+            "type_id": "ca.mcgill.a11y.image.renderer.Text",
+            "confidence": 50,
+            "description": renderingTitle + " (text only)",
+            "data": { "text": textString }
+        };
+        if (ajv.validate(textJSON, rendering["data"])) {
+            renderings.push(rendering);
+        } else {
+            console.error("Failed to generate a valid text rendering!");
+            console.error(ajv.errors);
+            console.warn("Trying to continue...");
+        }
+    } else {
+        console.debug("Skipped text rendering.");
+    }
+
+    // *******************************************************
+    // Haptic seg and obj coordinate data
+    // *******************************************************
+    const hapticSegInfo: { centroid: number[]; contourPoints: number[]; }[] = [];
+    const groupCentroidArray: Array<Array<number>> = [];
+    const groupCoordArray: Array<Array<number>> = [];
+
+    const segments = preSemSeg["segments"];

```

Comment:
This seems like it will fail if semantic segmentation didn't run/errored (which might happen!)